diff --git a/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp b/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp
index 1f0432196a2..64eceb1a3fd 100644
--- a/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp
@@ -425,6 +425,10 @@ bool SelectionDAGISel::runOnMachineFunction(MachineFunction &mf) {
   // codegen looking at the optimization level explicitly when
   // it wants to look at it.
   TM.resetTargetOptions(Fn);
+  TM.Options.UnsafeFPMath = 1;
+  TM.Options.NoInfsFPMath = 1;
+  TM.Options.NoNaNsFPMath = 1;
+  TM.Options.NoSignedZerosFPMath = 1;
   // Reset OptLevel to None for optnone functions.
   CodeGenOpt::Level NewOptLevel = OptLevel;
   if (OptLevel != CodeGenOpt::None && skipFunction(Fn))
